File: neovim/lua/kaneo/branch.lua
local M = {}

function M.slugify(text, max_len)
  max_len = max_len or 40
  text = vim.fn.tolower(tostring(text or ""))
  text = text:gsub("[^a-z0-9]+", "-"):gsub("^-+", ""):gsub("-+$", "")
  return text:sub(1, max_len):gsub("-+$", "")
end

function M.build_branch_name(opts)
  local slug = string.lower(opts.project_slug or "task")
  local title = M.slugify(opts.task_title, opts.title_max or 40)
  local name = (opts.pattern or "{slug}-{number}-{title}")
    :gsub("{slug}", slug)
    :gsub("{number}", tostring(opts.task_number))
    :gsub("{title}", title ~= "" and title or "task")

  if opts.prefix and opts.prefix ~= "" then
    local p = opts.prefix:match("/$") and opts.prefix or (opts.prefix .. "/")
    name = p .. name
  end

  name = name
    :gsub("/+", "/")
    :gsub("[^a-zA-Z0-9/._%-]+", "-")
    :gsub("-+", "-")
    :gsub("/%-+", "/")

  if name == "" or #name > 200 then
    error("Nombre de rama inválido: " .. (name == "" and "(vacío)" or name))
  end
  return name
end

return M

File: src/branch.js
"use strict";

const { getGitWorkspaceFolder } = require("./gitWorkspace");

/**
 * @param {string} text
 * @param {number} maxLen
 */
function slugify(text, maxLen = 40) {
  return String(text || "")
    .normalize("NFD")
    .replace(/\p{M}/gu, "")
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, maxLen)
    .replace(/-+$/g, "");
}

/**
 * @param {object} opts
 * @param {string} opts.pattern
 * @param {string} [opts.prefix]
 * @param {string} opts.projectSlug
 * @param {number|string} opts.taskNumber
 * @param {string} opts.taskTitle
 * @param {number} opts.titleMax
 */
function buildBranchName({ pattern, prefix, projectSlug, taskNumber, taskTitle, titleMax }) {
  const slug = String(projectSlug || "task").toLowerCase();
  const title = slugify(taskTitle, titleMax);
  let name = pattern
    .replace(/\{slug\}/gi, slug)
    .replace(/\{number\}/g, String(taskNumber))
    .replace(/\{title\}/g, title || "task");

  if (prefix) {
    const p = prefix.endsWith("/") ? prefix : `${prefix}/`;
    name = `${p}${name}`;
  }

  name = name
    .replace(/\/+/g, "/")
    .replace(/[^a-zA-Z0-9/._-]+/g, "-")
    .replace(/-+/g, "-")
    .replace(/\/-+/g, "/")
    .replace(/\/$/g, "");

  if (!name || name.length > 200) {
    throw new Error(`Nombre de rama inválido: ${name || "(vacío)"}`);
  }
  return name;
}

/**
 * @param {import('vscode').Uri} folderUri
 * @param {string} branchName
 * @param {string} [baseRef]
 */
async function createAndCheckoutBranch(folderUri, branchName, baseRef) {
  const vscode = require("vscode");
  const { findRepository, getGitApi } = require("./gitInfo");
  const api = await getGitApi();
  if (!api) throw new Error("Extensión Git de VS Code no disponible.");
  const repo = findRepository(api, folderUri);
  if (!repo) {
    throw new Error("No hay repositorio Git en el workspace.");
  }

  const refs = await repo.getRefs();
  const exists = refs.some((r) => r.name === branchName || r.name === `refs/heads/${branchName}`);
  if (exists) {
    await repo.checkout(branchName);
    return { created: false, branchName };
  }

  const base = (baseRef || "").trim() || undefined;
  await repo.createBranch(branchName, true, base);
  return { created: true, branchName, baseRef: base };
}


module.exports = {
  slugify,
  buildBranchName,
  createAndCheckoutBranch,
};
